utf8proc.c

```diff
@@ -685,22 +684,28 @@ UTF8PROC_DLLEXPORT utf8proc_ssize_t utf8proc_normalize_utf32(utf8proc_int32_t *b
         if (!starter_property) {
           starter_property = unsafe_get_property(*starter);
         }
-        if (starter_property->comb_index < 0x8000 &&
-            current_property->comb_index != UINT16_MAX &&
-            current_property->comb_index >= 0x8000) {
-          int sidx = starter_property->comb_index;
-          int idx = current_property->comb_index & 0x3FFF;
-          if (idx >= utf8proc_combinations[sidx] && idx <= utf8proc_combinations[sidx + 1] ) {
-            idx += sidx + 2 - utf8proc_combinations[sidx];
-            if (current_property->comb_index & 0x4000) {
-              composition = (utf8proc_combinations[idx] << 16) | utf8proc_combinations[idx+1];
-            } else
-              composition = utf8proc_combinations[idx];
-
-            if (composition > 0 && (!(options & UTF8PROC_STABLE) ||
-                !(unsafe_get_property(composition)->comp_exclusion))) {
-              *starter = composition;
-              starter_property = NULL;
+        int idx = starter_property->comb_index;
+        if (idx < 0x3FF && current_property->comb_issecond) {
+          int len = starter_property->comb_length;
+          utf8proc_uint32_t max_second = utf8proc_combinations[idx + len - 1][0];
+          if (current_char <= max_second) {
+            // TODO: binary search? arithmetic search?
+            for (int off = 0; off < len; ++off) {
+              utf8proc_uint32_t second = utf8proc_combinations[idx + off][0];
+              if (current_char < second) {
+                /* not found */
```

Comment:
Ideally this should be the fast path, since in most cases the second char will not be a combining character.

We may even want a special-case optimization for the common case of ASCII characters, since I think the second character of a combining pair can never be ASCII.

Reply:
I'm a little confused by the `current_char < second` branch, actually.  If `utf8proc_combinations[idx:idx+len-1]` is sorted in increasing order, wouldn't the "not found" case be `current_char > second`?

Reply:
Oh, I guess ASCII characters will never have `current_property->comb_issecond` set.